File: isegoria/src/main/java/com/eulersbridge/isegoria/friends/UserViewHolder.java
package com.eulersbridge.isegoria.friends;

import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.eulersbridge.isegoria.GlideApp;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.network.api.models.GenericUser;
import com.eulersbridge.isegoria.network.api.models.Institution;

class UserViewHolder extends RecyclerView.ViewHolder {

    interface OnClickListener {
        void onViewClick(@Nullable GenericUser user);
        void onActionClick(@Nullable GenericUser user);
    }

    private @Nullable GenericUser item;
    final private @Nullable OnClickListener clickListener;

    final private ImageView imageView;
    final private TextView nameTextView;
    final private TextView institutionTextView;

    UserViewHolder(View itemView, @DrawableRes int actionImageDrawableRes, @Nullable OnClickListener clickListener) {
        super(itemView);

        this.clickListener = clickListener;

        itemView.setOnClickListener(view -> {
            if (this.clickListener != null)
                this.clickListener.onViewClick(item);
        });

        imageView = itemView.findViewById(R.id.friends_list_image_view);
        nameTextView = itemView.findViewById(R.id.friends_list_name_text_view);
        institutionTextView = itemView.findViewById(R.id.friends_list_institution_text_view);

        ImageView actionImageView = itemView.findViewById(R.id.friends_list_action_image_view);
        actionImageView.setImageResource(actionImageDrawableRes);
        actionImageView.setOnClickListener(view -> {
            if (this.clickListener != null)
                this.clickListener.onActionClick(item);
        });
    }

    void setItem(@Nullable GenericUser item) {
        this.item = item;

        if (item == null) {
            nameTextView.setText(null);
            institutionTextView.setText(null);

        } else {
            nameTextView.setText(item.getFullName());

            GlideApp.with(imageView.getContext())
                    .load(item.profilePhotoURL)
                    .placeholder(R.color.white)
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(imageView);
        }
    }

    void setInstitution(@NonNull Institution institution) {
        if (item != null && item.institutionId != null && institution.id == item.institutionId)
            institutionTextView.setText(institution.getName());
    }
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/friends/UserViewHolder.kt
package com.eulersbridge.isegoria.friends

import android.support.annotation.DrawableRes
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.bumptech.glide.load.resource.bitmap.CircleCrop

import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.eulersbridge.isegoria.GlideApp
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.network.api.model.GenericUser

class UserViewHolder(
        itemView: View,
        private val clickListener: OnClickListener?
) : RecyclerView.ViewHolder(itemView) {

    private lateinit var item: GenericUser
    private lateinit var actionIcon: ActionIcon

    private val imageView: ImageView = itemView.findViewById(R.id.friends_list_image_view)
    private val nameTextView: TextView = itemView.findViewById(R.id.friends_list_name_text_view)
    private val subtextTextView: TextView= itemView.findViewById(R.id.friends_list_subtext_text_view)

    enum class ActionIcon {
        FRIEND,
        FRIEND_PENDING,
        ADD_FRIEND
    }

    interface OnClickListener {
        fun onViewClick(user: GenericUser?)
        fun onActionClick(user: GenericUser?)
    }

    init {
        itemView.setOnClickListener {
            clickListener?.onViewClick(item)
        }

        val actionImageView = itemView.findViewById<ImageView>(R.id.friends_list_action_image_view)
        actionImageView.apply {
            setOnClickListener {
                clickListener?.onActionClick(item)
            }
        }
    }

    fun setItem(item: GenericUser, actionIcon: ActionIcon) {
        this.item = item
        this.actionIcon = actionIcon

        bindItem(item, actionIcon)
    }

    private fun bindItem(user: GenericUser, actionIcon: ActionIcon) {
        if (user == null) {
            nameTextView.text = null
            subtextTextView.text = null

        } else {
            nameTextView.text = user.fullName
            subtextTextView.text = user.email

            GlideApp.with(imageView.context)
                    .load(user.profilePhotoURL)
                    .transform(CircleCrop())
                    .placeholder(R.drawable.account_circle_24dp)
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(imageView)

            // check to see if the person is already a friend
            // if they are then show already a friend icon
            // if theyre not show the add friend button

            val actionImageView = itemView.findViewById<ImageView>(R.id.friends_list_action_image_view)

            actionImageView.setImageResource(
                when (actionIcon) {
                    ActionIcon.FRIEND -> R.drawable.added_active
                    ActionIcon.FRIEND_PENDING -> R.drawable.added_inactive
                    ActionIcon.ADD_FRIEND -> R.drawable.add_friend_blue
                }
            )
        }
    }
}
